e2e-tests/testdata/echo_client/js/client.js

```diff
@@ -552,6 +552,178 @@ function mustBeSet(field, value) {
 }
 
 
+function encodeWebSocketHeaders(headers) {
+    // url safe, no pad
+    const base64encoded = btoa(JSON.stringify(headers))
+      .replaceAll("=", "")
+      .replaceAll("+", "-")
+      .replaceAll("/", "_");
+    return "encore.dev.headers." + base64encoded;
+}
+
+class WebSocketConnection {
+    hasUpdateHandlers = [];
+
+    constructor(url, headers) {
+        let protocols = ["encore-ws"];
+        if (headers) {
+            protocols.push(encodeWebSocketHeaders(headers));
+        }
+
+        this.ws = new WebSocket(url, protocols);
+
+        ws.addEventListener("error", () => {
+            this.resolveHasUpdateHandlers();
+        });
+
+        ws.addEventListener("message", () => {
+            this.resolveHasUpdateHandlers();
+        });
+
+        ws.addEventListener("close", () => {
+            this.resolveHasUpdateHandlers();
+        });
+    }
+
+    resolveHasUpdateHandlers() {
+        const handlers = this.hasUpdateHandlers;
+        this.hasUpdateHandlers = [];
+
+        for (const handler of handlers) {
+            handler()
+        }
+    }
+
+    async hasUpdate() {
+        // await until a new message have been received, or the socket is closed
+        await new Promise((resolve) => {
+            this.hasUpdateHandlers.push(() => resolve(null))
+        });
+    }
+
+    on(type, handler) {
+        this.ws.addEventListener(type, handler);
+    }
+
+    off(type, handler) {
+        this.ws.removeEventListener(type, handler);
+    }
+
+    close() {
+        this.ws.close();
+    }
+}
+
+export class BidiStream {
+    buffer = [];
+
+    constructor(url, headers) {
+        this.socket = new WebSocketConnection(url, headers);
+        this.socket.on("message", (event) => {
+            this.buffer.push(JSON.parse(event.data));
+        });
+    }
+
+    close() {
+        this.socket.close();
+    }
+
+    async send(msg) {
+        if (this.socket.ws.readyState === WebSocket.CONNECTING) {
+            // await that the socket is opened
+            await new Promise((resolve) => {
+                const handler = () => {
+                    this.socket.ws.removeEventListener("open", handler);
```

Comment:
Looks like there's an option to register the event listener only once: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#once

Reply:
fixed